lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,273 @@
+/**
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/* globals Logger */
+/* eslint-env browser */
+
+class ReportFeatures {
+
+  /**
+   * @param {!Document} document
+   */
+  constructor(document) {
+    this.json = null;
+    this._document = document;
+    this._copyAttempt = false;
+
+    this.onCopy = this.onCopy.bind(this);
+    this.onExportButtonClick = this.onExportButtonClick.bind(this);
+    this.onExport = this.onExport.bind(this);
+    this.onKeyDown = this.onKeyDown.bind(this);
+    this.printShortCutDetect = this.printShortCutDetect.bind(this);
+
+    this.logger = new Logger(this._document.querySelector('#lighthouse-log'));
+  }
+
+  _addEventListeners() {
+    this._setUpCollaspeDetailsAfterPrinting();
+
+    this.exportButton = this._document.querySelector('.lighthouse-export__button');
+    if (this.exportButton) {
+      this.exportButton.addEventListener('click', this.onExportButtonClick);
+      const dropdown = this._document.querySelector('.lighthouse-export__dropdown');
+      dropdown.addEventListener('click', this.onExport);
+
+      this._document.addEventListener('copy', this.onCopy);
+    }
+    this._document.addEventListener('keydown', this.printShortCutDetect);
+  }
+
+  /**
+   * Adds export button and print functionality to the report.
+   * @param {ReportJSON=} report
+   */
+  attach(report) {
+    this.json = report;
+    this._addEventListeners();
+  }
+
+  /**
+   * Handler copy events.
+   */
+  onCopy(e) {
+    // Only handle copy button presses (e.g. ignore the user copying page text).
+    if (this._copyAttempt) {
+      // We want to write our own data to the clipboard, not the user's text selection.
+      e.preventDefault();
+      e.clipboardData.setData('text/plain', JSON.stringify(this.json, null, 2));
+      this.logger.log('Report JSON copied to clipboard');
+    }
+
+    this._copyAttempt = false;
+  }
+
+  /**
+   * Copies the report JSON to the clipboard (if supported by the browser).
+   */
+  onCopyButtonClick() {
+    if (window.ga) {
+      window.ga('send', 'event', 'report', 'copy');
```

Comment:
Would love to keep GA for viewer. Should be fine since this is a noop for contexts that don't load ga.